feat(client): improve perfomance of generating path for client

sterligov opened the pull request:
In this pull request, I made a small performance improvement to the client by using string concatenation instead of fmt.Sprintf. Below, I’ll include the benchmark results.

```
goos: darwin
goarch: arm64
pkg: github.com/oapi-codegen/oapi-codegen/v2/pkg/codegen
cpu: Apple M1 Pro
BenchmarkFmtPath1param
BenchmarkFmtPath1param-10       	21433833	        49.99 ns/op	      16 B/op	       1 allocs/op
BenchmarkConcatPath1param
BenchmarkConcatPath1param-10    	73644105	        16.27 ns/op	       0 B/op	       0 allocs/op
BenchmarkFmtPath2param
BenchmarkFmtPath2param-10       	18985250	        62.78 ns/op	      16 B/op	       1 allocs/op
BenchmarkConcatPath2param
BenchmarkConcatPath2param-10    	55796110	        21.45 ns/op	       0 B/op	       0 allocs/op
BenchmarkFmtPath3param
BenchmarkFmtPath3param-10       	15205623	        79.05 ns/op	      24 B/op	       1 allocs/op
BenchmarkConcatPath3param
BenchmarkConcatPath3param-10    	41759646	        28.44 ns/op	       0 B/op	       0 allocs/op
PASS
```

Commit:
feat(client): improve perfomance of generating path for client

jamietanna commented:
Thanks, will try and review at some point. Is it worth committing the benchmark code for future testing?

Commit:
add benchmark code for concat path

sterligov commented:
> Thanks, will try and review at some point. Is it worth committing the benchmark code for future testing?

I added a commit with a benchmark test

mromaszewicz committed:
Merge remote-tracking branch 'upstream/main' into pr/2046

# Conflicts:
#	examples/generate/serverurls/gen.go
#	internal/test/any_of/codegen/ref_schema/openapi.gen.go
#	internal/test/any_of/param/param.gen.go
#	internal/test/issues/issue-1180/issue.gen.go
#	internal/test/issues/issue-1298/issue1298.gen.go
#	internal/test/issues/issue-1397/issue1397.gen.go
#	internal/test/issues/issue-grab_import_names/issue.gen.go
#	internal/test/options/response_getters/enabled/response_getters.gen.go
#	internal/test/references/external/pathalias/client.gen.go
#	internal/test/references/external/response_refs/pkg2/pkg2.gen.go
#	internal/test/references/external/x_go_name_qualification/gen/spec_base/issue.gen.go
#	internal/test/schemas/schemas.gen.go
#	internal/test/strict-server/client/client.gen.go

mromaszewicz committed:
chore: regenerate after merging main

Regenerates all generated code so the new string-concatenation path
builder in client.tmpl is applied to the test and example files that
moved during the "refactor all tests by category" reorganization.

mromaszewicz committed:
fix(codegen): handle percent signs and quotes in client paths

GenPathString built a format string and ran it through fmt.Sprintf at
generation time, so any percent sign in a spec path was interpreted as a
format verb. Percent-encoded paths such as /search/%20/{id} are legal
OpenAPI and produced corrupt output like

    "/search/%!/(string=  " + pathParam0 + ")%!s(MISSING)"

Deriving the parameter count from strings.Count(tmpl, "%s") compounded
this: a path containing a literal "%s" was counted as a parameter, so
/literal%s/{id} emitted a reference to an undeclared pathParam1.

Literal segments were also pasted into the generated source unquoted,
so a path containing a double quote or a backslash produced Go code
that did not compile or escaped incorrectly.

Build the concatenation from the regexp match offsets instead, quoting
each literal segment with strconv.Quote. Percent signs, quotes and
backslashes are then correct by construction, and the parameter count
comes from the number of matches rather than from counting needles in
an intermediate string.

Output is unchanged for every spec in the repository; `make generate`
produces no diff.

mromaszewicz committed:
chore(codegen): name the template helper genPathString

The surrounding template helpers are genParamArgs, genParamTypes and
genParamNames, and the Go function backing this one is GenPathString,
so getPathString reads as a typo. Rename it before it settles.

Generated output is unchanged.

mromaszewicz commented:
@greptileai review

mromaszewicz commented on pkg/codegen/template_helpers.go:
```diff
@@ -430,7 +430,7 @@ var TemplateFunctions = template.FuncMap{
 	"genParamTypes":              genParamTypes,
 	"genParamNames":              genParamNames,
-	"genParamFmtString":          ReplacePathParamsWithStr,
+	"genPathString":              GenPathString,
```
This is intentional, and I accept the consequences.

mromaszewicz committed:
fix(codegen): reuse the same variable for a repeated path parameter

A path parameter may appear more than once in one path, as in the
Keycloak-style /admin/realms/{realm}/clients/{client-uuid}/roles/{role-name}/composites/clients/{client-uuid}
that motivated #2220. SortParamsByPath deduplicates such parameters and
the client template declares one variable per deduplicated parameter,
but GenPathString numbered variables by occurrence, so the second
{client-uuid} referred to a pathParam that was never declared and the
generated client failed to compile.

Key the variable index by parameter name so repeats resolve back to the
first occurrence, and treat the prefixed and exploded forms {.param} and
{;param*} as the same parameter, matching OrderedParamsFromUri.

This also corrects the output rather than only restoring compilation.
The previous fmt.Sprintf form emitted one %s per occurrence but only one
argument per unique parameter, so the generated code failed go vet and
put a literal %!s(MISSING) in the request path at runtime.

mromaszewicz commented:
I found a few real bugs (and then so did Greptile) when reviewing your PR. I've pushed fixes and I'll merge this, so you get PR credit. Thanks for doing this. It's always good to avoid unnecessary function calls, even if the gain is minimal when you look at the big picture, vs a synthetic benchmark.
